File: clients/cloud/java/src/main/java/io/confluent/examples/clients/ConsumerExample.java
/**
 * Copyright 2015 Confluent Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.confluent.examples.clients;

import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Properties;
import java.util.Random;

public class ConsumerExample {
  public static void main(String[] args) throws Exception {
    if (args.length != 2) {
      System.out.println("Please provide command line arguments: configPath topic");
      System.exit(-1);
    }

    String topic = args[1];

    // Load properties from disk.
    Properties props = loadConfig(args[0]);

    // Add additional properties.
    props.put("enable.auto.commit", "true");
    props.put("auto.commit.interval.ms", "1000");
    props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
    props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");

    // WARNING: These settings are for CLI tools / examples. Don't use in production unless you want to reset the consumer app every reboot.
    props.put("group.id", "example" + new Random().nextInt());
    props.put("auto.offset.reset", "earliest");

    final Consumer<String, String> consumer = new KafkaConsumer<String, String>(props);
    consumer.subscribe(Arrays.asList(topic));

    try {
      while (true) {
        ConsumerRecords<String, String> records = consumer.poll(100);
        for (ConsumerRecord<String, String> record : records) {
          System.out.printf("offset = %d, key = %s, value = %s \n", record.offset(), record.key(), record.value());
        }
      }
    } finally {
      consumer.close();
    }
  }


  public static Properties loadConfig(String configFile) throws IOException {
    if (!Files.exists(Paths.get(configFile))) {
      throw new IOException(configFile + " not found.");
    }
    final Properties cfg = new Properties();
    try (InputStream inputStream = new FileInputStream(configFile)) {
      cfg.load(inputStream);
    }
    return cfg;
  }

}

File: clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/ConsumerExample.kt
/**
 * Copyright 2020 Confluent Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:JvmName("ConsumerExample")

package io.confluent.examples.clients.cloud

import io.confluent.examples.clients.cloud.model.DataRecord
import io.confluent.examples.clients.cloud.util.loadConfig
import io.confluent.kafka.serializers.KafkaJsonDeserializer
import io.confluent.kafka.serializers.KafkaJsonDeserializerConfig.JSON_VALUE_TYPE
import org.apache.kafka.clients.consumer.ConsumerConfig.*
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.serialization.StringDeserializer
import java.time.Duration.ofMillis
import kotlin.system.exitProcess

fun main(args: Array<String>) {

  if (args.size != 2) {
    println("Please provide command line arguments: configPath topic")
    exitProcess(1)
  }

  val topic = args[1]

  // Load properties from disk.
  val props = loadConfig(args[0])

  // Add additional properties.
  props[KEY_DESERIALIZER_CLASS_CONFIG] = StringDeserializer::class.java.name
  props[VALUE_DESERIALIZER_CLASS_CONFIG] = KafkaJsonDeserializer::class.java.name
  props[JSON_VALUE_TYPE] = DataRecord::class.java
  props[GROUP_ID_CONFIG] = "kotlin_example_group_1"
  props[AUTO_OFFSET_RESET_CONFIG] = "earliest"

  val consumer = KafkaConsumer<String, DataRecord>(props).apply {
    subscribe(listOf(topic))
  }

  var totalCount = 0L

  consumer.use {
    while (true) {
      totalCount = consumer
          .poll(ofMillis(100))
          .fold(totalCount, { accumulator, record ->
            val newCount = accumulator + 1
            println("Consumed record with key ${record.key()} and value ${record.value()}, and updated total count to $newCount")
            newCount
          })
    }
  }
}

